### DISCUSSION TITLE
Support for Shelly Plus RGBW PM

### DISCUSSION THREAD

**Discussion by PabloIT** [mongoose-os-apps/shelly-homekit#1502] [Ideas]
Will the new Shelly Plus RGBW PM be supported by Mongoose?
https://www.shelly.com/en/products/shop/shelly-plus-rgbw-pm

**Comment by markirb** [comment]
Oh wow did not know these exist.

If you or someone else here is up to donate the Hardware I would give it a try.

**Comment by nerdosity** [comment]
Got 2x. How can I help? 

![image](https://github.com/user-attachments/assets/5e4aaf87-40cf-4a96-b071-b6a6e135f7ce)
![image](https://github.com/user-attachments/assets/ab0c5f10-d1d0-43b8-9d6a-e1bf446263f8)
![image](https://github.com/user-attachments/assets/f9967bfd-4cfa-4252-9945-7b9e7d995b98)
![image](https://github.com/user-attachments/assets/c4fe6f16-2eda-4d58-9db3-c972a79c7d66)
![image](https://github.com/user-attachments/assets/c45c26ea-e213-4335-be34-731353ce8a1a)

**Comment by markirb** [comment]
Actually the pinout of the ESP32 would be needed. On the usual sites there is nothing yet...

**Comment by nerdosity** [comment]
> Actually the pinout of the ESP32 would be needed. On the usual sites there is nothing yet...

From left to right, screw adapters on the downside:
+12/24v GND GPIO36 GPIO37 GPIO38 GPIO39 GPIO25 GPIO26 GPIO27 GPIO24
So should be Input 1 to 4, R,G,B,W.

I did find everything with my multimeter.

Cannot find the onboard button yet, need a better magnifying lens.

**Comment by markirb** [comment]
Oh wow very nice. I can start to work with that.... 

I can compile rpc-service-gpio (https://mongoose-os.com/docs/mongoose-os/api/rpc/rpc-service-gpio.md) in for the first version then you could check on the device if you want.

**Comment by markirb** [comment]
Can you try to flash this? (cannot guarantee that it works, but normally this should not brick anything...)

Outputs should work like you expect, if you measured correctly :)

[shelly-homekit-ShellyPlusRGBWPM.zip](https://github.com/user-attachments/files/16319568/shelly-homekit-ShellyPlusRGBWPM.zip)

You can also play around with GPIOs like this:

curl 192.168.188.56/rpc/GPIO.Write -d '{"pin":1,"value":1}'

curl 192.168.188.56/rpc/GPIO.Read-d '{"pin":1}'

https://mongoose-os.com/docs/mongoose-os/api/rpc/rpc-service-gpio.md

this way you can press reset button and see what happens or try to find the LED output

**Comment by nerdosity** [comment]
I Will try, it's flashable via the UI?

**Comment by markirb** [comment]
Yes it should be.

**Comment by PabloIT** [comment]
@nerdosity Did you manage to test something?

**Comment by lanmarc77** [comment]
Is someone able to determine what the maximum input frequency is that could be measured by the inputs? I assume the inputs are using a low pass filter at least it looks like it based from the pictures.

**Comment by markirb** [comment]
@nerdosity any news?

@lanmarc77 do not know, what is the use case?

**Comment by lanmarc77** [comment]
@markirb I actually would not see a LED strip controller but a FAN controller. But that would require the inputs to be reasonable fast to read the FAN ticks to allow PI(D) control of the FAN speed.

**Comment by lanmarc77** [comment]
I can now answer the questions myself. I empirically determined that at around 300Hz (50% duty cycle) the rc filter cuts off. Removing a capacitor from one input I was able to detect until 15kHz and did not test further. 
So with this Shelly model FAN control is possible. For slower FANs without HW modification for faster FANs removal of the capacitor is needed.

**Comment by lanmarc77** [comment]
I measured the following pinout which almost matches @nerdosity 
    IN1: GPIO36
    IN2: GPIO37
    IN3: GPIO38
    IN4: GPIO39
    R: GPIO25
    G: GPIO26
    B: GPIO27
    W: GPIO4 (it is pin 24 not GPIO24)
    Button: GPIO22
    LED: GPIO14
    Power measurement: GPIO35 (OUT of unknown opamp circuit from measuring R001)
    Voltage monitor: GPIO34 (via unkown resistor network)

And a picture of the PCB back:
![IMG_20240825_205733_947](https://github.com/user-attachments/assets/6ccbfc7e-2fa7-4b33-bc95-91e2de1dd626)

I think there is a PCB temperature sensor somewhere but not yet find it.

**Comment by markirb** [comment]
Ok nice work. I will try to integrate this during the week and give you a test build.

**Comment by lanmarc77** [comment]
I have completely overwritten the Shelly. So I need a full setup with all initial ota, partiton table, bootloader....
This is what the Shelly partition layout looked like when I initially connected the serial adapter:
```
I (28) boot: Shelly OS loader 1.0.0 (Feb 25 2024 16:37:33 IDF 4.4.4-a8)
I (28) boot: chip revision: v3.1
I (29) boot_comm: chip revision: 3, min. bootloader chip revision: 0
I (36) boot.esp32: SPI Speed      : 80MHz
I (41) boot.esp32: SPI Mode       : DIO
I (45) boot.esp32: SPI Flash Size : 4MB
I (50) boot: Enabling RNG early entropy source...
I (55) boot: Partition Table @ 0x8000 (8ef27910):
I (60) boot: ## Label            Usage          Type ST Offset   Length   F
I (68) boot:  0 nvs              WiFi data        01 02 00009000 00004000 0
I (76) boot:  1 otadata          OTA data         01 00 0000d000 00002000 1
I (83) boot:  2 app_0            OTA app          00 10 00010000 00190000 1
I (91) boot:  3 fs_0             FS               01 82 001a0000 00060000 1
I (99) boot:  4 app_1            OTA app          00 11 00200000 00190000 1
I (106) boot:  5 fs_1             FS               01 82 00390000 00060000 1
I (114) boot:  6 aux              unknown          55 00 003f0000 0000c000 0
I (122) boot:  7 shelly           Unknown data     01 88 003fc000 00004000 1
I (130) boot: End of partition table
I (134) ota: Read BS0: seq 14 bst 0x11 seq2 14 | as 1 rs 0 ba 2 mfs 0 c 1 | v 1
I (142) ota: Read BS1: seq 14 bst 0x11 seq2 14 | as 1 rs 0 ba 2 mfs 0 c 1 | v 1
I (150) ota: Booting app 1
I (153) boot_comm: chip revision: 3, min. application chip revision: 0
I (160) esp_image: segment 0: paddr=00200020 vaddr=3f400020 size=2e75ch (190300) map
I (226) esp_image: segment 1: paddr=0022e784 vaddr=3ff80000 size=00004h (     4) load
I (227) esp_image: segment 2: paddr=0022e790 vaddr=3ffbdb60 size=01888h (  6280) load
I (234) esp_image: segment 3: paddr=00230020 vaddr=400d0020 size=13f624h (1308196) map
I (636) esp_image: segment 4: paddr=0036f64c vaddr=3ffbf3e8 size=0214ch (  8524) load
I (639) esp_image: segment 5: paddr=003717a0 vaddr=40080000 size=1d818h (120856) load
I (698) boot: Loaded app from partition at offset 0x200000
I (698) boot: Disabling RNG early entropy source...
cpu_start: Unicore app
```

**Comment by markirb** [comment]
Try this:

[shelly-homekit-ShellyPlusRGBWPM.zip](https://github.com/user-attachments/files/16747269/shelly-homekit-ShellyPlusRGBWPM.zip)

can you give me feedback then if everything works?
especially if the configuration is retained which means the button is correctly mapped and not inverted.

you have to flash via mos-tool

**Comment by lanmarc77** [comment]
The back button is inverted. When I press it the serial console reads zero if I release 1, followed by a reset procedure.
Setting the color values results in core dumps. I attached one for you to check. I think at least red channel works as far as I could test.
Onboard LED seems to be ok. It is on during bootup then switches off. If I keep the backbutton pressed it flashes.
How would I test the inputs?

[core-PlusRGBWPM-esp32-20240826-134901.594212025.zip](https://github.com/user-attachments/files/16748189/core-PlusRGBWPM-esp32-20240826-134901.594212025.zip)

**Comment by markirb** [comment]
> The back button is inverted. When I press it the serial console reads zero if I release 1, followed by a reset procedure.

This is expected behavior. So nothing to change


> Setting the color values results in core dumps. I attached one for you to check. 

Strange. I need some time to analyze this. I think it would be best to test in 4White mode so you can test each individual channel.


> I think at least red channel works as far as I could test. 

OK

> Onboard LED seems to be ok. It is on during bootup then switches off. If I keep the backbutton pressed it flashes. How would I test the inputs?

Good question. You can via above mentioned GPIO-RPC for example.
Otherwise, currently they are not really connected in the firmware :/
In 4 white mode we could have one switch for each white? It is not currently the case


> 
> [core-PlusRGBWPM-esp32-20240826-134901.594212025.zip](https://github.com/user-attachments/files/16748189/core-PlusRGBWPM-esp32-20240826-134901.594212025.zip)

**Comment by lanmarc77** [comment]
Ok. I was able to verify the RGB channels. They all work correctly. Also the inputs. They deliver 1 if pulled to GND, 0 otherwise.
Is the white channel not supported?

**Comment by markirb** [comment]
It should be? Maybe Pin 24 all along?

**Comment by lanmarc77** [comment]
I meant how do I actually set it up? The webinterface only offers setting rgb but this model has an additional white channel.

**Comment by lanmarc77** [comment]
And I am pretty certain it is GPIO4=PIN24. I marked the signal path from the ESP32 through the gate driver to the corresponding mosfet and inserted the relevant pinouts from the datasheet:
![grafik](https://github.com/user-attachments/assets/951df50d-85a3-4fc2-8f89-1ff25c8e182a)

**Comment by markirb** [comment]
> I meant how do I actually set it up? The webinterface only offers setting rgb but this model has an additional white channel.

![Bildschirmfoto 2024-08-28 um 07 12 28](https://github.com/user-attachments/assets/abb40062-ba7f-45ea-be3f-1094f2256231)

**Comment by lanmarc77** [comment]
Ah Ok. Thats how it looks on my device:
![Screenshot_20240828-130651](https://github.com/user-attachments/assets/07a16b44-f4d4-47c4-94f9-50c94e55bbfb)


no Mode dropdown.

**Comment by markirb** [comment]
Ah. My bad. Will send you a new version

**Comment by markirb** [comment]
[shelly-homekit-ShellyPlusRGBWPM.zip](https://github.com/user-attachments/files/16787900/shelly-homekit-ShellyPlusRGBWPM.zip)

this should work

**Comment by lanmarc77** [comment]
That worked. White is outputting data and changes it according to what I see in the serial log. I guess we are done then?

**Comment by markirb** [comment]
Ok thanks nice. 

I will release this as alpha along the mini versions in the next days when I find time.

What's missing is PM support but that would take a while.
